lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -263,22 +263,34 @@ def add_credit_card(xml, credit_card, options = {})
           xml.tag! 'CVNum', credit_card.verification_value if credit_card.verification_value?
 
           if options[:three_d_secure]
-            three_d_secure = options[:three_d_secure]
-            xml.tag! 'BuyerAuthResult' do
-              xml.tag! 'Status', three_d_secure[:status] unless three_d_secure[:status].blank?
-              xml.tag! 'AuthenticationId', three_d_secure[:authentication_id] unless three_d_secure[:authentication_id].blank?
-              xml.tag! 'PAReq', three_d_secure[:pareq] unless three_d_secure[:pareq].blank?
-              xml.tag! 'ACSUrl', three_d_secure[:acs_url] unless three_d_secure[:acs_url].blank?
-              xml.tag! 'ECI', three_d_secure[:eci] unless three_d_secure[:eci].blank?
-              xml.tag! 'CAVV', three_d_secure[:cavv] unless three_d_secure[:cavv].blank?
-              xml.tag! 'XID', three_d_secure[:xid] unless three_d_secure[:xid].blank?
-            end
+            add_three_d_secure(options[:three_d_secure], xml)
           end
 
           xml.tag! 'ExtData', 'Name' => 'LASTNAME', 'Value' =>  credit_card.last_name
         end
       end
 
+      def add_three_d_secure(three_d_secure, xml)
+        xml.tag! 'BuyerAuthResult' do
+          xml.tag! 'Status', three_d_secure[:status] unless three_d_secure[:status].blank?
+          xml.tag! 'AuthenticationId', three_d_secure[:authentication_id] unless three_d_secure[:authentication_id].blank?
+          xml.tag! 'AUTHENTICATION_STATUS', authentication_status(three_d_secure)
+          xml.tag! 'PAReq', three_d_secure[:pareq] unless three_d_secure[:pareq].blank?
+          xml.tag! 'ACSUrl', three_d_secure[:acs_url] unless three_d_secure[:acs_url].blank?
+          xml.tag! 'ECI', three_d_secure[:eci] unless three_d_secure[:eci].blank?
+          xml.tag! 'CAVV', three_d_secure[:cavv] unless three_d_secure[:cavv].blank?
+          xml.tag! 'XID', three_d_secure[:xid] unless three_d_secure[:xid].blank?
+        end
+
+        xml.tag! 'ExtData', 'Name' => 'THREEDSVERSION', 'Value' =>  three_d_secure[:version] unless three_d_secure[:version].blank?
+        xml.tag! 'ExtData', 'Name' => 'DSTRANSACTIONID', 'Value' =>  three_d_secure[:ds_transaction_id] unless three_d_secure[:ds_transaction_id].blank?
+      end
+
+      def authentication_status(three_d_secure)
```

Comment:
this method is returning `nil` if not values are present. we could either replace it or change it to something like
```ruby
three_d_secure[:authentication_response_status] || three_d_secure[:directory_response_status]
```